File: scala/bob/src/main/scala/bob.scala
class Bob {
  def isSilence(str: String) = str.forall(_.isWhitespace)
  def isQuestion(str: String) = str.endsWith("?")
  def isYell(str: String) = str.exists(_.isUpper) && !str.exists(_.isLower)

  def hey(str: String): String = str match {
    case s if isSilence(s) => "Fine. Be that way!"
    case s if isYell(s) => "Whoa, chill out!"
    case s if isQuestion(s) => "Sure."
    case _ => "Whatever."
  }
}

File: haskell/bob/Bob.hs
module Bob (responseFor) where

import Data.Char (isLetter, isSpace, toUpper)

responseFor :: String -> String
responseFor sentence
    | isSilence = "Fine. Be that way!"
    | isShout = "Whoa, chill out!"
    | isQuestion = "Sure."    
    | otherwise = "Whatever." where
        isSilence = all isSpace sentence
        isQuestion = last sentence == '?'
        isUpper = map toUpper sentence == sentence
        isShout = isUpper && containsLetter
        containsLetter = any isLetter sentence        